<a href="https://colab.research.google.com/github/pawan-97/FDS_JOURNAL/blob/FDS--LAB/FDS_LAB_Practical_9_HypothesisTesting_AB_Testing_PawanKumar.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive 
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Required libraries

```
# This is formatted as code
```



In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math
import scipy.stats as stats
from scipy.stats import ttest_1samp
from statsmodels.stats.power import tt_ind_solve_power

# Hypothesis Testing or Significance Testing

Hypothesis testing is a statistical method that is used in making statistical decisions using experimental data. Hypothesis Testing is basically an assumption that we make about the population parameter.

There are many different statistical significance, or hypothesis, tests. They all follow the same basic principle. The appropriate test for a given situation depends on the nature of the data being analysed.

Null hypothesis:

All statistical significance tests start with a null hypothesis. A statistical significance test measures the strength of evidence that the data sample supplies for or against some proposition of interest.

This proposition is known as a 'null hypothesis', since it usually relates to there being 'no difference' between groups' or 'no effect' of a treatment.

For example:

1.CMV infected babies have the same average birthweight as non-infected babies.

2.ß-thalassaemia does not have any effect on ferritin level.

3.Recombinant human erythropoietin has no effect on the haemoglobin levels in premature infants.

Even if our hypothesis is not to do with 'no difference' it is still convention that the hypothesis being tested is known as the null hypothesis

---
---

**Example 1:**

In [ ]:
data=pd.read_excel('/content/drive/MyDrive/BMI.xls')
data

Name  Age  Gender  Height  Weight   BMI Value   BMI Status
0  Shivani    21      F     164       60       22.3       Normal
1   Gaurav    22      M     180       83       25.6   Overweight
2    Hamza    25      M     176       65       21.0       Normal
3   Preeti    26      F     160       62       24.2       Normal
4   Aniket    30      M     182       90       27.2   Overweight
5    Suraj    27      M     170       45       15.6  Underweight
6    Nidhi    18      F     158       72       28.8   Overweight
7    Kayum    20      M     170       55       19.0  Underweight

1) check if population mean value of BMI is significantly different than 23 . Use 0.05 as α (los).

Null Hypothesis will be

Ho : mu = 23

Alternate Hypothesis will be

H1 : mu != 23

# One Sample t-test
This test is appropriate when a continuous variable is recorded on a single sample of individuals or items and the mean value is to be tested against some hypothesised population mean

Here we use t test as we don't know the population variance and the sample size is less than 30.

In [ ]:
def hypothesisTesting(df,column,H0,H1,alpha,mean,tailTest):
    ts, pv = ttest_1samp(df[column].values, mean)
    t2pv = pv
    t1pv = pv*2
    if tailTest == 1:
        if t1pv < alpha:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
    else:
        if t2pv < alpha/2:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
 
hypothesisTesting(data,'BMI Value','mean = 23','mean != 23',0.05,23,2)

Null Hypothesis: Not Rejected
Conclusion: mean = 23


From above result we can say that population mean of BMI is equal to 23.

---
---

### Example 2:

2) check if population mean value of Height is greater than 170 cm . Use 0.05 as α (los).

Null Hypothesis will be

Ho : mu <= 170

Alternate Hypothesis will be

H1 : mu > 170

Here we use t test as we know the poplation variance and the sample size is less than 30.

In [ ]:
def hypothesisTesting(df,column,H0,H1,alpha,mean,tailTest):
    ts, pv = ttest_1samp(df[column].values, mean)
    t2pv = pv
    t1pv = pv*2
    if tailTest == 1:
        if t1pv < alpha:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
    else:
        if t2pv < alpha/2:
            print("Null Hypothesis: Rejected")
            print("Conclusion:",H1)
        else:
            print("Null Hypothesis: Not Rejected")
            print("Conclusion:",H0)
 
hypothesisTesting(data,'Height','mean <= 170','mean > 170',0.05,170,1)

Null Hypothesis: Not Rejected
Conclusion: mean <= 170


From above result we can say that population mean of Height is less or equal to 170 cm.

### Example **3**:

In [ ]:
df=pd.read_excel('/content/drive/MyDrive/result.xlsx')
df

Roll no.     Name  Subj1  Subj2  subj3  total result
0       101   Manoj      45     45     45    135      P
1       102   Manish     35     45     42    122      P
2       103  shubham     29     26     30     85      P
3       104   shivam     38     35     29    102      P
4       105   aakash     41     40     34    115      P
5       106   ritwik     46     62     41    149      P
6       107  saurabh     29     48     27    104      P
7       108   nilesh     43     33     33    109      P
8       109  prakash     37     30     38    105      P
9       110  vikrant     33     31     41    105      P

3) Check if total mean value of marks is not more than 113.

Null Hypothesis will be

Ho : mu <= 113

Alternate Hypothesis will be

Ha : mu > 113

In [ ]:
Ho = "mu <= 113"
# alt hyp
Ha = "mu > 113"
# alpha
al = 0.05
# mu - mean
mu = 113
# tail type
tt = 1
# data
marks = df['total'].values
print("Ho:", Ho)
print("Ha:", Ha)
print("al:", al)
print("mu:", mu)
print(marks)
print("")
ts, pv = ttest_1samp(marks, mu)
print("t-stat",ts)
print("p-vals",pv)
t2pv = pv
t1pv = pv*2
print("1t pv",t1pv)
print("2t pv",t2pv)

if tt == 1:
    if t1pv < al:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)
else:
    if t2pv < al/2:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)

Ho: mu <= 113
Ha: mu > 113
al: 0.05
mu: 113
[135 122  85 102 115 149 104 109 105 105]

t-stat 0.01733524930528476
p-vals 0.9865473848679749
1t pv 1.9730947697359498
2t pv 0.9865473848679749
Null Hypothesis: Not Rejected
Conclusion: mu <= 113


## Example 4:

In [ ]:
iris=pd.read_excel("/content/drive/MyDrive/iris-m.xlsx")
iris

Id  SepalLength  SepalWidth  PetalLength  PetalWidth           species
0      1          5.1         3.5          1.4         0.2       Iris-setosa
1      2          4.9         3.0          1.4         0.2       Iris-setosa
2      3          4.7         3.2          1.3         0.2       Iris-setosa
3      4          4.6         3.1          1.5         0.2       Iris-setosa
4      5          5.0         3.6          1.4         0.2       Iris-setosa
..   ...          ...         ...          ...         ...               ...
145  146          6.7         3.0          5.2         2.3    Iris-virginica
146  147          6.3         2.5          5.0         1.9    Iris-virginica
147  148          6.5         3.0          5.2         2.0    Iris-virginica
148  149          6.2         3.4          5.4         2.3  Iris-virginica  
149  150          5.9         3.0          5.1         1.8  Iris-virginica  

[150 rows x 6 columns]

4) Check if the population mean value
of PetalWidth is no more than 1.5 .


Null Hypothesis will be

Ho : mu <= 1.5

Alternate Hypothesis will be

H1 : mu > 1.5


In [ ]:
Ho = "mu <= 1.5"

Ha = "mu > 1.5"
# alpha
al = 0.05
# mu - mean
mu = 1.5
# tail type
tt = 1
# data
petalwidth= iris['PetalWidth'].values

# print
print("Ho:", Ho)
print("Ha:", Ha)
print("al:", al)
print("mu:", mu)
print(petalwidth)
print("")

ts, pv = ttest_1samp(petalwidth, mu)
print("t-stat",ts)
print("p-vals",pv)
t2pv = pv
t1pv = pv*2
print("1t pv",t1pv)
print("2t pv",t2pv)

if tt == 1:
    if t1pv < al:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)
else:
    if t2pv < al/2:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)

Ho: mu <= 1.5
Ha: mu > 1.5
al: 0.05
mu: 1.5
[0.2 0.2 0.2 0.2 0.2 0.4 0.3 0.2 0.2 0.1 0.2 0.2 0.1 0.1 0.2 0.4 0.  0.3
 0.3 0.3 0.2 0.4 0.2 0.5 0.2 0.2 0.4 0.2 0.2 0.2 0.2 0.4 0.1 0.2 0.1 0.2
 0.2 0.1 0.2 0.2 0.3 0.3 0.2 0.6 0.4 0.3 0.2 0.2 0.2 0.2 1.4 1.5 1.5 1.3
 1.5 1.3 1.6 1.  1.3 1.4 1.  1.5 1.  1.4 1.3 1.4 1.5 1.  1.5 1.1 1.8 1.3
 1.5 1.2 1.3 1.4 1.4 1.7 1.5 1.  1.1 1.  1.2 1.6 1.5 1.6 1.5 1.3 1.3 1.3
 1.2 1.4 1.2 1.  1.3 1.2 1.3 1.3 1.1 1.3 2.5 1.9 2.1 1.8 2.2 2.1 1.7 1.8
 1.8 2.5 nan 1.9 2.1 2.  2.4 2.3 1.8 2.2 2.3 1.5 2.3 2.  2.  1.8 2.1 1.8
 1.8 1.8 2.1 1.6 1.9 2.  2.2 1.5 1.4 2.3 2.4 1.8 1.8 2.1 2.4 2.3 1.9 2.3
 2.5 2.3 1.9 2.  2.3 1.8]

t-stat nan
p-vals nan
1t pv nan
2t pv nan
Null Hypothesis: Not Rejected
Conclusion: mu <= 1.5


The population mean value
of PetalWidth is equal to 6.25.


### **Example 5:**

5) Use relevant statistical test to check if population mean value of SepalLength is significantly different from 6.25.

Null Hypothesis will be 

H0 : mu = 6.25

Alternate Hypothesis will be

H1 : mu != 6.25

In [ ]:
#null hyp
Ho = "mu = 6.25"
# alt hyp
Ha = "mu != 6.25"
# alpha
al = 0.06
# mu - mean
mu = 6.25
# tail type
tt = 2

# data
sepallength = iris['SepalLength'].values
# print
print("Ho:", Ho)
print("Ha:", Ha)
print("al:", al)
print("mu:", mu)
print(sepallength)
print("")

ts, pv = ttest_1samp(sepallength, mu)
print("t-stat",ts)
print("p-vals",pv)
t2pv = pv
t1pv = pv*2
print("1t pv",t1pv)
print("2t pv",t2pv)

if tt == 1:
    if t1pv < al:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)
else:
    if t2pv < al/2:
        print("Null Hypothesis: Rejected")
        print("Conclusion:",Ha)
    else:
        print("Null Hypothesis: Not Rejected")
        print("Conclusion:",Ho)

Ho: mu = 6.25
Ha: mu != 6.25
al: 0.06
mu: 6.25
[5.1 4.9 4.7 4.6 5.  0.  4.6 5.  4.4 4.9 5.4 4.8 4.8 4.3 5.8 5.7 5.4 5.1
 5.7 5.1 5.4 5.1 4.6 5.1 4.8 5.  5.  5.2 5.2 4.7 4.8 5.4 5.2 5.5 4.9 5.
 5.5 4.9 4.4 5.1 5.  4.5 4.4 5.  5.1 4.8 5.1 4.6 5.3 nan 7.  6.4 6.9 5.5
 6.5 5.7 6.3 4.9 6.6 5.2 5.  5.9 6.  6.1 5.6 6.7 5.6 5.8 6.2 5.6 5.9 6.1
 6.3 6.1 6.4 6.6 6.8 6.7 6.  5.7 5.5 5.5 5.8 6.  5.4 6.  6.7 6.3 5.6 5.5
 5.5 6.1 5.8 5.  5.6 5.7 5.7 6.2 5.1 5.7 6.3 5.8 7.1 6.3 6.5 7.6 4.9 7.3
 6.7 7.2 6.5 6.4 6.8 5.7 5.8 6.4 6.5 7.7 7.7 6.  6.9 5.6 7.7 6.3 6.7 7.2
 6.2 6.1 6.4 7.2 7.4 7.9 6.4 6.3 6.1 7.7 6.3 6.4 6.  6.9 6.7 6.9 5.8 6.8
 6.7 6.7 6.3 6.5 6.2 5.9]

t-stat nan
p-vals nan
1t pv nan
2t pv nan
Null Hypothesis: Not Rejected
Conclusion: mu = 6.25




 The Population mean value of SepalLength is equal to 6.25.



---
---

# AB Testing

Hypothesis testing is a broader term that encompasses a/b testing. Typically a hypothesis test will have a null hypothesis which could be tested against two or more samples for significance.

A/B Testing goes a step further and tries to calculate implied impact from the experiment testing two randomized different variants. Randomization helps t o ensure bias is minimized but this isn't always easy to accomplish.

Example :

# Two Sample t test

# Example 1:

First, we’ll create two arrays to hold the measurements of each group of 20 plants:

In [ ]:
group1 = np.array([14, 15, 15, 16, 13, 8, 14, 17, 16, 14, 19, 20, 21, 15, 15, 16, 16, 13, 14, 12])
group2 = np.array([15, 17, 14, 17, 14, 8, 12, 19, 19, 14, 17, 22, 24, 16, 13, 16, 13, 18, 15, 13])

Before we perform the test, we need to decide if we’ll assume the two populations have equal variances or not. As a rule of thumb, we can assume the populations have equal variances if the ratio of the larger sample variance to the smaller sample variance is less than 4:1.

In [ ]:
#find variance for each group
print(np.var(group1), np.var(group2))

7.727500000000001 12.260000000000002


The ratio of the larger sample variance to the smaller sample variance is 12.26 / 7.73 = 1.586, which is less than 4. This means we can assume that the population variances are equal.

Thus, we can proceed to perform the two sample t-test with equal variances:

The two hypotheses for this particular two sample t-test are as follows:

H0: µ1 = µ2 (the two population means are equal)

HA: µ1 ≠µ2 (the two population means are not equal)

In [ ]:
#perform two sample t-test with equal variances
stats.ttest_ind(a=group1, b=group2, equal_var=True)

Ttest_indResult(statistic=-0.6337397070250238, pvalue=0.5300471010405257)

Because the p-value of our test (0.53005) is greater than alpha = 0.05, we fail to reject the null hypothesis of the test.

We do not have sufficient evidence to say that the mean height of plants between the two populations is different.

# Type I error

A type 1 error is also known as a false positive and occurs when a researcher incorrectly rejects a true null hypothesis. This means that your report that your findings are significant when in fact they have occurred by chance.

The probability of making a type I error is represented by your alpha level (α), which is the p-value below which you reject the null hypothesis. A p-value of 0.05 indicates that you are willing to accept a 5% chance that you are wrong when you reject the null hypothesis.

You can reduce your risk of committing a type I error by using a lower value for p. For example, a p-value of 0.01 would mean there is a 1% chance of committing a Type I error.

However, using a lower value for alpha means that you will be less likely to detect a true difference if one really exists (thus risking a type II error).

# Example 1:

In above eg , the p value is 0.53005 which is considered as type I error.

# Type II error

A type II error is also known as a false negative and occurs when a researcher fails to reject a null hypothesis which is really false. Here a researcher concludes there is not a significant effect, when actually there really is.

The probability of making a type II error is called Beta (β), and this is related to the power of the statistical test (power = 1- β). You can decrease your risk of committing a type II error by ensuring your test has enough power.

You can do this by ensuring your sample size is large enough to detect a practical difference when one truly exists.

Taking example from two sample t test 

In [ ]:
effect_size=abs((group1.mean()-group2.mean())/(group1.std()-group2.std()))
sample_size=20
alpha=0.05
ratio=1.0

statistical_power = tt_ind_solve_power(effect_size=effect_size, nobs1=sample_size, alpha=alpha, ratio=1.0, alternative='two-sided')
print(statistical_power)

0.7927312603881692


In [ ]:
# type 2 error 
type_2_error=1-statistical_power
type_2_error


0.2072687396118308

Proability of Type 2 error is 0.2073